### PR TITLE
add login feature

### PR DESCRIPTION
**Pull request by Harshtroo**

### CODE REVIEW

File: tms/clickup/constant.py [hunk 1]
```diff
@@ -1 +1,2 @@
 PASSWORD_ERROR_MESSAGE = "Password not Match."
+LOGIN_ERROR_MESSAGE = "Invalid Username and Password."
```
Comment (Dhruvi-Makwana) [inline tms/clickup/constant.py]: "enter valid username or password"

File: tms/clickup/serializer.py [hunk 2]
```diff
@@ -22,3 +23,17 @@ def create(self, validated_data):
         validated_data.pop('password_confirm')
         user = super().create(validated_data)
         return user
+
+
+class LoginSerializer(serializers.Serializer):
+    username = serializers.CharField()
+    password = serializers.CharField()
+
+    def create(self, validated_data):
```
Comment (Dhruvi-Makwana) [inline tms/clickup/serializer.py]: we don't need create method.

File: tms/clickup/serializer.py [hunk 3]
```diff
@@ -22,3 +23,17 @@ def create(self, validated_data):
...
+        username = validated_data["username"]
+        password = validated_data["password"]
+        user = authenticate(username=username, password=password)
+        if not user:
+            raise serializers.ValidationError(constant.LOGIN_ERROR_MESSAGE)
+        token, _ = Token.objects.get_or_create(user=user)
```
Comment (Dhruvi-Makwana) [inline tms/clickup/serializer.py]: do not genrate token here , if user is authenticate then return user else error that's it.

File: tms/clickup/views.py [hunk 4]
```diff
@@ -1,7 +1,19 @@
...
 class SingUpView(CreateAPIView):
     serializer_class = RegistrationSerializer
+
+
+class LoginView(APIView):
+    def post(self, request):
+        serializer = LoginSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        token = serializer.save()
+        return JsonResponse({'token': token.key,
```
Comment (Dhruvi-Makwana) [inline tms/clickup/views.py:19]: here pass user with token

File: tms/clickup/serializer.py [hunk 5]
```diff
@@ -29,11 +29,16 @@ class LoginSerializer(serializers.Serializer):
...
+        password = data.get('password')
+
+        if username and password:
+            user = authenticate(username=username, password=password)
+            if not user:
+                raise serializers.ValidationError("Incorrect Username and Password. Please try again.")
+        else:
+            raise serializers.ValidationError("Both username and password are required.")
+
+        data['user'] = user
```
Comment (Dhruvi-Makwana) [inline tms/clickup/serializer.py]: don't need data[user] simply return  user

      "  return user "

File: tms/clickup/serializer.py [hunk 6]
```diff
@@ -29,11 +29,16 @@ class LoginSerializer(serializers.Serializer):
...
+    def validate(self, data):
+        username = data.get('username')
+        password = data.get('password')
+
+        if username and password:
+            user = authenticate(username=username, password=password)
+            if not user:
+                raise serializers.ValidationError("Incorrect Username and Password. Please try again.")
+        else:
+            raise serializers.ValidationError("Both username and password are required.")
```
Comment (Dhruvi-Makwana) [inline tms/clickup/serializer.py]: no need to add else conditions. 
if not auth user raise error 
if not user:
                raise serializers.ValidationError("Incorrect Username and Password. Please try again.")
return user

**Review by yatishTT — APPROVED** [review]

**Pull request merged**